### DISCUSSION TITLE
TypeScript Docs for _document.tsx

### DISCUSSION THREAD

**Discussion by bnhr** [ben-rogerson/twin.macro#429] [Q&A]
Is there any specific config for _document.tsx?
I faced this issue:
`Property 'ids' does not exist on type 'Readonly<RenderPageResult & { styles?: ReactFragment | ReactElement<any, string | JSXElementConstructor<any>>[]; } & { ...; }> & Readonly<...>'.`
and
`Property 'css' does not exist on type 'Readonly<RenderPageResult & { styles?: ReactFragment | ReactElement<any, string | JSXElementConstructor<any>>[]; } & { ...; }> & Readonly<...>'.ts(2339)`

**Comment by ben-rogerson** [comment]
Have you checked out [these next.js docs](https://nextjs.org/docs/advanced-features/custom-document#typescript) yet?

**Reply by bnhr** [reply]
Yes, but for not for this data-emotion
` <style data-emotion-css={this.props.ids.join(' ')} dangerouslySetInnerHTML={{ __html: this.props.css }}/>`

**Reply by ben-rogerson** [reply]
Could you please show us how to recreate this issue?

**Comment by rottitime** [comment]
I have created a fix. @ben-rogerson can you please approve https://github.com/ben-rogerson/twin.examples/pull/105. The fix provides better types for `ctx`

```lang-js
//_document.tsx
import Document, {
  Html,
  Head,
  Main,
  NextScript,
  DocumentContext,
  DocumentInitialProps
} from 'next/document'
import { extractCritical } from '@emotion/server'

type NewDocumentInitialProps = DocumentInitialProps & {
  ids: string[]
  css: string
}

class CustomDocument extends Document<NewDocumentInitialProps> {
  static async getInitialProps(ctx: DocumentContext): Promise<DocumentInitialProps> {
    const initialProps = await Document.getInitialProps(ctx)
    const critical = extractCritical(initialProps.html)
    initialProps.html = critical.html
    initialProps.styles = (
      <>
        {initialProps.styles}
        <style
          data-emotion-css={critical.ids.join(' ')}
          dangerouslySetInnerHTML={{ __html: critical.css }}
        />
      </>
    )

    return initialProps
  }

  render() {
    return (
      <Html lang="en">
        <Head>
          <style
            data-emotion-css={this.props?.ids?.join(' ')}
            dangerouslySetInnerHTML={{ __html: this.props.css }}
          />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default CustomDocument
```
@ben-rogerson Would it be ok to add this for Typescript support please?
